wordpress-mobile/WordPress-FluxC-Android
# Publish to s3

This PR adds support for publishing artifacts to S3 through a set of Buildkite jobs.

Due to the modularized nature of this library, there was a major decision to be made on how to setup the publishing flow. Specifically, if we were to directly publish the library, we'd end up with `unspecified` version of dependencies that are added as `project`s. Gradle does not (directly) support a fat-aar approach and there isn't a built-in way to deal with these multi-module projects within `maven-publish` plugin. Here are the 2 possible ways to address this issue:

1. We can publish artifacts all together, but update the generated POM file immediately before publishing and manually set the version of `project` dependencies.
2. We can publish artifacts subsequently, starting with modules with no `project` dependencies and work our way up by replacing `project` dependencies with the published binaries.

The trade-offs of first approach:

* It's much simpler with the main complication being isolated to the `pom` property within `publishing` block
* It's faster because all artifacts can be published at the same time
* This is essentially a hack and does not represent how the project was built at the time of publishing
* There is no guarantee that the modules are published correctly

The advantages of second approach:

* Each module is built the way it'll be imported to the projects (using the binary dependency) - Technically, the resolved dependency _can_ be different, but that's always true and besides the point
* It guarantees that the internal modules are published "correctly". For example, while publishing `fluxc-processor` we'll fetch `fluxc-annotations` from S3 and build the project with that
* It's slower to publish
* Binary dependency substitution to publish creeps into the `dependencies` section of each project

---

The second approach is a clear winner for me because of the publishing guarantees. The only thing I am not particularly happy about is the dependency substitution creeping into the build files. That's why I explored using [`resolutionStrategy`](https://docs.gradle.org/current/userguide/resolution_rules.html) to keep it contained, however to get that to work we'd have to add the resolved dependencies to POM instead of the declared dependencies and I am not sure of its impacts. I also couldn't get it to work the way I wanted and even then it wasn't as contained as I wanted it to be.

_P.S: Note that this approach is the one we used in [gutenberg-mobile](https://github.com/wordpress-mobile/gutenberg-mobile) project, so by picking second approach, we are also keeping things consistent across projects._

The manual dependency substitution worked out better than I expected with just 3 conditional properties defined in root `build.gradle` extension: `fluxcAnnotationsProjectDependency`, `fluxcProcessorProjectDependency`, `fluxcProjectDependency`

---

For publishing, I opted for separate scripts for each module and they are pretty straightforward. They first read the published version of the other modules and then pass them to the project as Gradle properties.

**To test:**
* Smoke test WPAndroid & WCAndroid: 
https://github.com/wordpress-mobile/WordPress-Android/pull/15119 
https://github.com/woocommerce/woocommerce-android/pull/4547
* Verify that composite build works for WPAndroid & WCAndroid
